sb3_contrib/ppo_mask/ppo_mask.py

```
@@ -0,0 +1,438 @@
+    """
+
+    def __init__(
+        self,
+        policy: Union[str, Type[MaskableActorCriticPolicy]],
+        env: Union[GymEnv, str],
+        learning_rate: Union[float, Schedule] = 3e-4,
+        n_steps: int = 2048,
+        batch_size: Optional[int] = 64,
+        n_epochs: int = 10,
+        gamma: float = 0.99,
+        gae_lambda: float = 0.95,
+        clip_range: Union[float, Schedule] = 0.2,
+        clip_range_vf: Union[None, float, Schedule] = None,
+        ent_coef: float = 0.0,
+        vf_coef: float = 0.5,
+        max_grad_norm: float = 0.5,
+        target_kl: Optional[float] = None,
+        tensorboard_log: Optional[str] = None,
+        create_eval_env: bool = False,
+        policy_kwargs: Optional[Dict[str, Any]] = None,
+        verbose: int = 0,
+        seed: Optional[int] = None,
+        device: Union[th.device, str] = "auto",
+        _init_setup_model: bool = True,
+        use_masking: bool = True,
+    ):
+        super().__init__(
+            policy,
+            env,
+            learning_rate=learning_rate,
+            n_steps=n_steps,
+            gamma=gamma,
+            gae_lambda=gae_lambda,
+            ent_coef=ent_coef,
+            vf_coef=vf_coef,
+            max_grad_norm=max_grad_norm,
+            use_sde=False,
+            sde_sample_freq=-1,
+            tensorboard_log=tensorboard_log,
+            create_eval_env=create_eval_env,
+            policy_kwargs=policy_kwargs,
+            verbose=verbose,
+            seed=seed,
+            device=device,
+            _init_setup_model=False,
+            supported_action_spaces=(
+                spaces.Discrete,
+                spaces.MultiDiscrete,
+                spaces.MultiBinary,
+            ),
+        )
+
+        if use_masking and not is_masking_supported(self.env):
+            logger.warn("Environment does not support action masking. Consider using ActionMasker wrapper")
```

I would throw an error, no?

---

why would someone do that?
There is normal PPO for such env, no?

---

if the test env does not support it, it can still provide an invalid action method, no? (which will return a vector that says every action is valid)

---

I refactored so that `use_masking` is a kwarg for the `learn()` method, rather than a model attribute specified at construction time. That way `use_masking` isn't included in the saved model.

It's now an error if an env (either training or eval) fails to implement the masking API when `use_masking` is true.